README
==============
Sib
==============

Sib is a dictionary for Gnome. Dictonary is written in python ( English to Persian and Persian to Engilish ) The database have 50259 Engilish words from xfardic dictionary .

Feature

   1. Trayicon support
   2. Auto translate on selected word with mouse in any window and page
   3. Translate notify
   4. Auto startup at gnome ( Can disable from gnome-session-properties / Startup programs (tab) )
   5. with Database manager can add new or edit database or import BGL file 

Requirements

   1. pygtk (>= 2.10)
   2. python-sqlite2
   3. python-notify (>=0.1) 


Recommended

   1. festival


Basic Installation
==================

   1. make
   2. make install ( root user ) 


ChangeLog
Tuesday, March 03 2009 Mola Pahnadayan <mola.mp@gmail.com>
	* Fix if ~/.sib not exist
	* Set default of "Search on while" to enable
	
Monday, February 23 2009 Mola Pahnadayan <mola.mp@gmail.com>
	* Add database manager
	* import from BGL file ( babylon glossary library )
	* can change font of translate
	* New Icon
	* Festival added
	* Code clean and bug fix

Friday, August 29 2008 Foad Nosrati Habibi <foad.nh@gmail.com>
	*handle words ended to: ed, d, es, s, ing & ly
	
Friday, June 13 2008  Mola Pahnadayan  <mola.mp@gmail.com>
	* Add dialog of database manager
	* Some buf fix

Wednesday, Nov 14 2007 Milad Rastian <miladmovie@gmail.com>
	* Add espeak support 

Tuesday, May 22 2007  Mola Pahnadayan  <mola.mp@gmail.com>
	* Add authors file
	* Add Global hotkey
	* Bug fix



src/prefrence.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
##
#    
#    Copyright (C) 2007  Mola Pahnadayan
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
##

import pygtk
pygtk.require20()
import gtk
import gtk.glade
import os

Optionfile="/sib.conf"
Option = {'notifytimer':5000,
          'searchontype':True,
          'history':10,
          'Dfont':False,
          'Sfont':None}

def get_home():
    home = os.environ.get('HOME', '')
    return home+"/.sib"
    
def option_parse():
    home = get_home()
    if os.path.isfile(home+Optionfile):
        sibconf = open(home+Optionfile, 'r')
        for element in sibconf:
            sin,sout = element.split(' = ')
            Option[sin] = sout[:-1]
        sibconf.close()
                
class prefrence:
    def __init__(self,datadir,pixdir,dsearch,dtime,dhist,Dfont,Sfont):
        self.glade=gtk.glade.XML(datadir+"/interface.glade","Preference")
        self.dlg=self.glade.get_widget("Preference")
        dict = {"on_preference_close":self.cancel,
              "on_cancelbutton1_clicked":self.cancel,
              "on_okbutton1_clicked":self.close,
              "on_radiobutton_changed":self.groupchange,
              "on_fontselection1_set_focus_child":self.fontselection, 
              }
        self.glade.signal_autoconnect(dict)      
        
        self.spin1 = self.glade.get_widget("spinbutton1")
        self.spin2 = self.glade.get_widget("spinbutton2")
        self.checkbutton = self.glade.get_widget("checkbutton1")
        self.font_sel = self.glade.get_widget("fontselection1")
        self.checkbutton.set_active(dsearch)
        self.spin1.set_value(dtime/1000)
        self.spin2.set_value(dhist)
        
        self.Dfont=Dfont
        self.Sfont=Sfont
        if Dfont:
            self.glade.get_widget("radiobutton2").set_active(True)
            self.font_sel.set_sensitive(True)
            self.font_sel.set_font_name(Sfont)
            
        
        self.image= gtk.Image()
        self.image.set_from_file(pixdir+"/sib.png")
        
        if os.path.exists(get_home())==False:
            os.mkdir(get_home())
        

    def close(self, obj, data=None):
        global Option
        home = get_home()
        timer = self.spin1.get_value_as_int()*1000
        history = self.spin2.get_value_as_int()
        search = self.checkbutton.get_active()
        
        Option['searchontype']=search
        Option['notifytimer']=timer
        Option['history']=history
        Option['Dfont']=self.Dfont
        Option['Sfont']=self.Sfont

        sibconf=open(home+Optionfile,'w')
        sibconf.write('notifytimer = '+str(timer)+'\n')
        sibconf.write('history = '+str(history)+'\n')
        sibconf.write('searchontype = '+str(search)+'\n')
        sibconf.write('Dfont = '+str(self.Dfont)+'\n')
        sibconf.write('Sfont = '+str(self.Sfont)+'\n')
        sibconf.close()
        
        return 1
    
    def cancel(self, obj, data=None):
        self.dlg.destroy()
    
    def groupchange(self, obj, data=None):
        if obj.get_name()=="radiobutton2":
            self.font_sel.set_sensitive(True)
            self.Dfont =True
            self.Sfont = self.font_sel.get_font_name()
            return
        elif obj.get_name()=="radiobutton1":
            self.font_sel.set_sensitive(False)
            self.Dfont=False
    
    def fontselection(self, obj, data=None):
        self.Sfont = obj.get_font_name()


src/sib.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
##
#    
#    Copyright (C) 2007  Mola Pahnadayan
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
##

import pygtk
pygtk.require20()
import gtk, gobject
import gtk.glade, gtk.gdk
import os
import pynotify
import commands
import pango

import prefrence
import dbmanager
import dbquery

__DATADIR__ = "/usr/share/sib"
__PIXDIR__ = "/usr/share/pixmaps/sib"

class main:
    version = "0.8"
    tray = None
    clipboardtimer = 1000
    lastclipboard = ""
    lastword = ""
    reverse = False
    shownotify = False
    minimize = 0
    chistory = 0
    word_change = 0
    ## for Speech
    festival = 'echo "%s"|festival --tts'
    espeak = 'espeak "%s"'
          
    def clipboard_text_received(self, clipboard, text, data):
        if not text or text == '' or text[0] == ' ':
            return
        ch = ord(text[0])
        if not( (ch>=97 and ch<=122)or(ch>=65 and ch<=90)):
            return
        
        if self.lastclipboard != text:
            word = text
            word = self.parse(word)
            if self.lastword != word:
                self.shownotify = True
                self.comboentry.insert_text(0, word)
                self.comboentry.set_active(0)
                self.word_change = 0
                self.lastword = word

            self.chistory += 1
            if self.chistory>self.history:
                self.comboentry.remove_text(self.history)
                self.chistory -= 1
            self.lastclipboard = text
        return
    
    def fetch_clipboard_info(self):
        self.clipboard.request_text(self.clipboard_text_received)
        return True
        
    def activate_icon_cb(self, data=None):
        if self.minimize == 1:
            self.window.set_property("skip-taskbar-hint", False)
            self.window.deiconify()
            self.minimize = 0
            return
        
        if self.window.get_property('visible'):
            self.window.hide()
        else:
            self.window.set_property("skip-taskbar-hint", False)
            self.window.present()

    def popup_menu_cb(self, obj, button, time):
        menu = self.pmenu.get_widget("menu1")
        menu.popup(None, None, None, button, time)
    
    def TrayLoad(self):
        self.statusIcon = gtk.StatusIcon()
        self.statusIcon.set_from_file(__PIXDIR__+"/sib-24.png")
        self.statusIcon.set_tooltip("Dictonary Sib")
        self.statusIcon.connect('activate', self.activate_icon_cb)
        self.statusIcon.connect('popup-menu', self.popup_menu_cb)
        self.statusIcon.set_visible(True)

        self.clipboard = gtk.clipboard_get(gtk.gdk.SELECTION_PRIMARY)
        self.clipboard.request_text(self.clipboard_text_received)
        gobject.timeout_add(self.clipboardtimer, self.fetch_clipboard_info)
        
    def about_dialog(self, obj):
        if (self.about is None):
            self.about = gtk.AboutDialog()
            self.about.set_name("Sib is a dictionary")
            #self.about.set_version(self.version)
            self.about.set_authors(["Mola Pahnadayan  <mola.mp@gmail.com>", 
                                    "   Project Manager, Maintainer",
                                    "\nThanks to",
                                    "  Foad Nosrati Habibi  <foad.nh@gmail.com>"
                                    ])
            file = open(__DATADIR__+"/COPYING")
            license = file.read()
            file.close()
            self.about.set_license(license)
            pixmap = gtk.gdk.pixbuf_new_from_file(__PIXDIR__+"/about.png")
            self.about.set_logo(pixmap)
            self.about.set_transient_for(self.window)
            self.about.set_website("http://sib.ospdev.net")
            def response_cb(widget, id):
                return True
            self.about.connect ("response", response_cb)
            result=self.about.run()
            self.about.destroy()
            self.about = None
        
    def __init__(self):
        ghome=os.environ.get('HOME', '')+'/.sib'
        if os.path.exists(ghome)==False:
            os.mkdir(ghome)
        prefrence.option_parse()
        self.speak = self.festival
        self.dbs = dbmanager.lbase()
        self.glade = gtk.glade.XML(__DATADIR__+"/interface.glade","window1")
        self.window = gtk.Window()
        self.window = self.glade.get_widget("window1")
        dict = {"on_window1_delete_event":self.wdelete,
              "on_comboboxentry1_changed":self.change,
              "on_treeview1_cursor_changed":self.cursorchange,
              "on_togglebutton1_toggled":self.reversesearch,
              "on_window1_window_state_event":self.mainchange,
              "on_treeview1_button_press_event":self.popupcopy,
              "on_btn_speech_clicked":self.speechplay,
              }
        self.glade.signal_autoconnect(dict)      
        
        self.comboentry = gtk.ComboBoxEntry()
        self.comboentry = self.glade.get_widget("comboboxentry1")
        self.store = gtk.ListStore(str)
        self.comboentry.set_model(self.store)
        self.comboentry.set_text_column(0)
        self.entry1 = self.comboentry.get_child()
        self.entry1.connect("key_press_event",self.entry_keypress)

        ## Speech button
        self.btn_speech = gtk.Button()
        self.btn_speech = self.glade.get_widget("btn_speech")
                        
        toggle = self.glade.get_widget("togglebutton1")
        image = gtk.Image()
        image.set_from_file(__PIXDIR__+"/reverse.png")
        toggle.set_image(image)
        self.pmenu = gtk.glade.XML(__DATADIR__+"/interface.glade","menu1")
        mdict = {"on_quit1_activate":gtk.main_quit,
               "on_about1_activate":self.about_dialog,
               "on_preferences1_activate":self.loadpref,
               "on_DBmanager_activate":self.loadDBmanager
               }
        self.pmenu.signal_autoconnect(mdict)
        
        self.about = None
        self.DB_con = dbquery.Database(__DATADIR__)
        
        self.window.set_icon_from_file(__PIXDIR__+"/sib.png")
        self.TrayLoad()
        
        self.treestore = gtk.ListStore(str,str)

        self.treeview = self.glade.get_widget("treeview1")
        self.treeview.set_model(self.treestore)
        
        self.cell = gtk.CellRendererText()
        col = gtk.TreeViewColumn("Word",self.cell,text = 0)
        self.treeview.append_column(col)
        self.treeview.set_search_column(0)
        
        self.textv = self.glade.get_widget("textview1")
        self.btn_speech = self.glade.get_widget("btn_speech")
        if os.path.exists("/usr/bin/festival"):
            self.btn_speech.set_sensitive(True)
        self.option_applay()

        self.window.hide()
    
    def change(self, obj):
        if self.searchontype:
            id = self.comboentry.get_active_text()
            self.translate(id)
        self.word_change = 0


    def translate(self, id = None):
        string = None
        conv = 0
        if len(id)>0:
            if self.reverse:
                string = self.query('wname','wmean like "%% %s %%"'%(id))
            else:
                string = self.query('wname','wname like "%s%%"'%(id))
                while string == None and conv != 2:
                    id, conv = self.convert(id, conv)
                    string = self.query('wname','wname like "%s%%"'%(id))                
        self.treestore.clear()
        if string != None:
            for s in string:
                self.treestore.append([s,None])
            count = self.glade.get_widget("statusbar1")
            buff = " Find %d words" % len(string)
            count.push(0,buff)
        else:
            textview = self.glade.get_widget("textview1")
            buf = textview.get_buffer()
            buf.set_text("Word not find")
            textview.set_buffer(buf)
        row = self.treestore.get_iter_first()
        if row!= None:
            path = self.treestore.get_path(row)
            self.treeview.set_cursor(path, None, False)

    def convert(self, word, conv):
        if word[-1:] == 'd' and conv == 0:
            return word[:-1], 1
        elif word[-1:] == 'e' and conv == 1:
            return word[:-1], 2
        elif word[-1:] == 's' and conv == 0:
            return word[:-1], 1
        elif word[-3:] == 'ing' and conv == 0:
            return word[:-3], 2
        elif word[-2:] == 'ly' and conv == 0:
            return word[:-2], 2
        else:
            return word, 2

    def cursorchange(self, obj):
        selection = obj.get_selection()
        (mode,iter) = selection.get_selected()
        word = mode.get(iter,0)[0]
        string = self.query('wmean','wname = "%s"'%(word))
        if len(string)>0:
            text = ""
            for s in string:
                text += unicode(s) + ','
            string = text
        if not(self.window.is_active()) :
            self.notify(word,unicode(string))
        textview = self.glade.get_widget("textview1")
        buf = textview.get_buffer()
        buf.set_text(string)
        textview.set_buffer(buf)
        self.word_change = 1

    def query(self, field, state):
        find = self.DB_con.get_all(field, state)
        if len(find)!= 0 :
            return find
        
        return None
    
    def notify(self, word, msg):
        pynotify.init("Dictonary Sib")
        n = pynotify.Notification(word, msg, __PIXDIR__+"/sib.png")
        n.set_urgency(pynotify.URGENCY_NORMAL )
        n.set_timeout(self.notifytimer) 
        if not n.show():
            print "Can't create notify"
        
    def reversesearch(self, obj):
        self.reverse = obj.get_active()

    def mainchange(self, obj, state):
        event = state.new_window_state
        if event == gtk.gdk.WINDOW_STATE_ICONIFIED:
            self.window.set_property("skip-taskbar-hint", True)
            self.minimize = 1
            
    def wdelete(self, obj, event):
        self.window.hide()
        return True
    
    def parse(self, str):
        string = ""
        for char in str:
            ch = ord(char)
            if ( (ch>=97 and ch<=122)or(ch>=65 and ch<=90)):
                string+=char
            else:
                break
                
        return string

    def copytoclipboard(self, obj):
        selection=self.treeview.get_selection()
        (mode,iter) = selection.get_selected()
        if iter:
            word = mode.get(iter,0)[0]
            self.shownotify = False
            c = gtk.Clipboard()
            c.set_text(''.join(word))
            c.store()
        
    def popupcopy(self, button, event):
        if event.button==3:
            menu = gtk.Menu()
            mcopy = gtk.ImageMenuItem(stock_id=gtk.STOCK_COPY)
            mcopy.connect("activate",self.copytoclipboard)
            mcopy.show()
            menu.insert(mcopy,0)
            menu.popup(None, None, None, event.button, event.time)

    def entry_keypress(self, obj, key):
        if key.keyval==gtk.keysyms.Return:
            word = self.comboentry.get_active_text()
            self.shownotify = False
            self.comboentry.insert_text(0,word)
            self.lastword = word
            self.chistory += 1
            self.word_change = 0
            if self.chistory>self.history:
                self.comboentry.remove_text(self.history)
                self.chistory -= 1
            if not(self.searchontype):
                self.translate(word)

    def loadpref(self, obj, data=None):
        dialog = prefrence.prefrence(__DATADIR__,__PIXDIR__,self.searchontype,
                            self.notifytimer,self.history,self.Dfont,self.Sfont)
        feedback=dialog.dlg.run()
        dialog.dlg.destroy()
        self.option_applay()
    
    def loadDBmanager(self, obj, data=None):
        dialog = dbmanager.DBmanager(__DATADIR__)
            
    def option_applay(self):
        try:
            if str(prefrence.Option['searchontype'])[0] == 'T':
                self.searchontype = True
            else:
                self.searchontype = False
        except:
            self.searchontype = True

        try:
            self.notifytimer =int( prefrence.Option['notifytimer'])
        except:
            self.notifytimer = 5000
        
        try:
            self.history = int(prefrence.Option['history'])
        except:
            self.history = 10
        
        if str(prefrence.Option['Dfont'])[0]=='T':
            self.Dfont = True
        else:
            self.Dfont = False
        
        try:
            self.Sfont = str(prefrence.Option['Sfont'])
        except:
            self.Sfont = ""

        if (self.Dfont==True)and(self.Sfont!=None):
            self.textv.modify_font(pango.FontDescription(self.Sfont))
        else:
            self.textv.modify_font(pango.FontDescription(""))

    
    def speechplay(self,obj,data=None):
        if self.comboentry.get_active_text() == '':
            return
        if self.word_change == 0:
            out_put = commands.getstatusoutput(self.speak %(self.comboentry.get_active_text()))
        else:
            selection = self.treeview.get_selection()
            (mode,iter) = selection.get_selected()
            word = mode.get(iter,0)[0]
            out_put = commands.getstatusoutput(self.speak %(word))
            
        if out_put[1].find("Invalid device") >= 0:
            pynotify.init("Dictonary Sib")
            n = pynotify.Notification("Warnning", "Sound device is used by another program", __PIXDIR__+"/sib.png")
            n.set_urgency(pynotify.URGENCY_CRITICAL)
            n.set_timeout(self.notifytimer) 
            if not n.show():
                print "Can't create notify"
#            self.notify("Error", "Sound device is used by another program")
#            msgWindow=gtk.MessageDialog(None,0,gtk.MESSAGE_WARNING,gtk.BUTTONS_CLOSE,'Error')
#            msgWindow.format_secondary_text('Sound device is used by another program' )
#            response=msgWindow.run()
#            msgWindow.destroy()
        
if __name__ == "__main__":
    main()
    gtk.main()
